### PR TITLE
Add system to import functions from remote

### PR DESCRIPTION
**Pull request by pyrech**
For now, only import from GitHub repositories is supported.

```php
use function Castor\import;

import('github://<user>/<repository>/<path to php file to import>@<version>');
```

A trust system is embedded to avoid malicious remote execution and ask users if they allow to import remote functions.

image

See the documentation for how it works: https://github.com/jolicode/castor/blob/remote-imports/doc/15-remote.md

Future ideas:

- Make use of this feature in docker-starter to be able to provide optional infrastructure parts:
```php
import('github://jolicode/docker-starter/worker')
```
- Provide some "standard" library of common tasks we used at JoliCode:
```php
import('github://jolicode/docker-starter/v4/.castor/qa.php')
```

### CODE REVIEW

File: doc/13-remote.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+When importing functions from a remote resource, Castor will download the files
+and store them in `$HOME/.castor/remote/`.
+
+### From a GitHub repository
+
+To import functions from a GitHub repository, pass a path to the `import()`
+function, formatted like this:
+
+```
+github://<user>/<repository>/<version>/<path of the php file to import>
```
Comment (lyrixx) [inline doc/13-remote.md:16]: what about 
```suggestion
github://<user>/<repository>/<path of the php file to import>@<version>
```
?

File: src/GlobalHelper.php [hunk 2]
```diff
@@ -129,7 +130,28 @@ public static function getCache(): CacheItemPoolInterface&CacheInterface
...
+        return self::$homeDirectory ??= PlatformUtil::getUserDirectory();
+    }
+
+    /**
+     * @throws \RuntimeException If the user home could not reliably be determined
+     */
+    public static function getGlobalDirectory(): string
+    {
+        return self::getHomeDirectory() . '/.castor';
+    }
```
Comment (lyrixx) [inline src/GlobalHelper.php:156]: I would 

1. remove `getHomeDirectory`, it's not used
2. rename `setHomeDirectory` in `getGlobalDirectory`

File: src/PlatformUtil.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        if (\function_exists('posix_getuid') && \function_exists('posix_getpwuid')) {
+            $info = posix_getpwuid(posix_getuid());
+
+            if ($info) {
+                return $info['dir'];
+            }
+        }
+
+        throw new \RuntimeException('Could not determine user directory');
```
Comment (lyrixx) [inline src/PlatformUtil.php:48]:
```suggestion
        throw new \RuntimeException('Could not determine user directory.');
```

File: src/Remote/Exception/NotTrusted.php [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Castor\Remote\Exception;
+
+class NotTrusted extends \RuntimeException
+{
+    public function __construct(
+        public readonly string $url,
+        public readonly bool $wasAsked = true,
```
Comment (lyrixx) [inline src/Remote/Exception/NotTrusted.php:9]: not used?

File: src/Remote/Import.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            throw new NotTrusted($url, false);
+        }
+
+        $trustKey = sprintf('remote.trust.%s', str_replace('/', '.', $url));
+
+        $trustChoice = cache(
+            $trustKey,
+            function (ItemInterface $item) {
+                if ($item->isHit()) {
+                    return $item->get();
```
Comment (lyrixx) [inline src/Remote/Import.php:59]: this condidition always return false, right? we are in the path where it's a cache miss.

File: src/Remote/Import.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                }
+
+                $item->expiresAfter(-1);
```
Comment (lyrixx) [inline src/Remote/Import.php:62]: And by default item does't expire, so it's uslees?

File: src/Remote/Import.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+            if (TrustEnum::ALWAYS === $trustChoice) {
+                return;
+            }
+        }
+
+        static $displayTrustWarning = true;
+
+        if ($displayTrustWarning) {
+            $io->warning('Your Castor project tries to import functions from external resources');
```
Comment (lyrixx) [inline src/Remote/Import.php:83]:
```suggestion
            $io->warning('Your Castor project tries to import functions from external resources.');
```

File: tests/Examples/ArgsAnotherArgsTest.php [hunk 8]
```diff
@@ -9,7 +9,7 @@ class ArgsAnotherArgsTest extends TaskTestCase
     // args:another-args
     public function test(): void
     {
-        $process = $this->runTask(['args:another-args', 'FIXME(required)', '--test2', 1]);
+        $process = $this->runTask(['args:another-args', 'FIXME(required)', '--test2', 1, '--no-trust']);
```
Comment (lyrixx) [inline tests/Examples/ArgsAnotherArgsTest.php:12]: Is there a way to not update all tests?

**Review by lyrixx — COMMENTED** [review]
Very cool

File: tests/Examples/ArgsAnotherArgsTest.php (hunk 8 above)
Comment (pyrech) [reply tests/Examples/ArgsAnotherArgsTest.php:12]: Did not find a way, so I'm open to any suggestions

File: src/Remote/Import.php (hunk 5 above)
Comment (pyrech) [reply src/Remote/Import.php:59]: Nope, It's a bit hackish, but it's a workaround to be able to:
- read from the cache if the choice was persisted before
- write in the same cache key after asking to the user

Without that, the callable of the second cache call (line 102) below would never be called

File: src/Remote/Import.php (hunk 6 above)
Comment (pyrech) [reply src/Remote/Import.php:62]: Idem

File: src/Remote/Import.php (hunk 5 above)
Comment (pyrech) [reply src/Remote/Import.php]: Changed to use the cache directly

File: src/Remote/Import.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        self::ensureTrustedResource($domain . '/' . $repository);
+
+        $dir = GlobalHelper::getHomeDirectory() . '/remote/' . $domain . '/' . $repository . '/' . $version;
+
+        if (!is_dir($dir) && !$dryRun) {
+            log("Importing functions in path {$functionPath} from {$domain}/{$repository} (version {$version})");
+
+            fs()->mkdir($dir);
+
+            run(['git', 'clone', "git@{$domain}:{$repository}.git", '--branch', $version, '--depth', '1', '--filter', 'blob:none', '.'], path: $dir, quiet: true);
```
Comment (damienalexandre) [inline src/Remote/Import.php:55]: **❔ question:** ‏Is it possible to add a nice error message for server than don't have "git" by default? (most unix distributions).

Comment (pyrech) [reply src/Remote/Import.php:55]: This is ran on your host, so I guess most developers have git installed :upside_down_face:. Anyway, I guess we can add a check to error gracefully :+1:

Comment (pyrech) [reply src/Remote/Import.php]: Done

**Comment by TheoD02** [comment]
Hey,

Not a lot of movement here? ^^
What's the status of this PR? What's the reason for this long inactivity ? What's left to do?😄

I'm waiting for this PR to split the common code to share it with different projects 😁

**Comment by pyrech** [comment]
For the record we sometimes talk internally about this feature we all want. But we do not fully agree on some details, mainly where to store the downloaded files. I hope we can find a solution soon but not sure it will find its way before the release of the v1

**Comment by lyrixx** [comment]
We had a very good talk with @pyrech and @joelwurtz and here the ADR

The goal is to generate a `composer.json`, then use composer to download everything.

##  Default usage, package is published on packagist.org

```php
import('composer://jolicode/jolitypo');
```

Add this to composer.json
```json
{
    "require": {
        "jolicode/castor": "*",
    }
}
```

Then execute at runtime
```php
require_once PathHelper::getRoot() . '/.castor/vendor/autoload.php';
```

## More complex usage

```php
import('composer://jolicode/jolitypo',
    file: 'path/to/file.php',
    version: 'v1.2.34',
);
```

Add this to composer.json
```json
{
    "require": {
        "jolicode/castor": "v1.2.34",
    }
}
```

Then execute at runtime
```php
require_once PathHelper::getRoot() . '/.castor/vendor/autoload.php';
require_once PathHelper::getRoot() . '/.castor/vendor/jolicode/jolitypo/path/to/file.php';
```

## Package that is not published on packagist.org

```php
import('composer://jolicode/jolitypo',
    file: 'path/to/file.php',
    version: 'v1.2.34',
    vcs: 'https://github.com/jolicode/jolitypo.git',
);
```


Add this to composer.json
```json
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/jolicode/jolitypo.git"
        }
    ],
    "require": {
        "jolicode/castor": "v1.2.34",
    }
}
```

Then execute at runtime
```php
require_once PathHelper::getRoot() . '/.castor/vendor/autoload.php';
require_once PathHelper::getRoot() . '/.castor/vendor/jolicode/jolitypo/path/to/file.php';
```

## Not a package

```php
import('package://what/you-want',
    file: 'path/to/file.php',
    // version: 'dev-main', // Not allowed here
    source: [
        "url" =>  "https://github.com/soyuka/php-wasm.git",
        "type" => "git",
        "reference" => "main"
    ],
);
```

Add this to composer.json
```json
{
    "repositories": [
        {
            "type": "package",
            "package": {
                "name": "what/you-want",
                "version": "dev-main",
                "source": {
                    "url": "https://github.com/soyuka/php-wasm.git",
                    "type": "git",
                    "reference": "main"
                }
            }
        },

    ],
    "require": {
        "what/you-want": "*",
    }
}
```

Then execute at runtime
```php
require_once PathHelper::getRoot() . '/.castor/vendor/autoload.php';
```

**Comment by pyrech** [comment]
Closing in favor of #297

**Pull request closed without merging**